### ISSUE TITLE
[BUG] file:sync doesn't respect conf.xml serializer; hard codes serialization parameters

### ISSUE THREAD

**Issue by joewiz** [eXist-db/exist#3704]
**Describe the bug**

The `file:sync` function, which writes collection contents out to disk, hard codes serialization parameters such as `indent=yes`. This inserts whitespace into data and can pollute application-sensitive whitespace practices and lead to undesired source control changes.

**Expected behavior**

All functions that serialize data should, at bare minimum, respect the `<serializer>` settings in conf.xml. 

See https://github.com/eXist-db/exist/blob/develop/extensions/modules/file/src/main/java/org/exist/xquery/modules/file/Sync.java#L91-L96.

**To Reproduce**

The final expression below returns `false()` even when conf.xml's `serializer/@index` is set to `"no"`. If it respected this setting, the expression would return `true()`.

```xquery
xquery version "3.1";

xmldb:create-collection("/db", "test"),
xmldb:store("/db/test", "test.xml", <foo><bar/></foo>),
file:mkdir("/Users/joe/test"),
file:sync("/db/test", "/Users/joe/test", ()),
deep-equal(<foo><bar/></foo>, doc("file:/Users/joe/test/test.xml"))
```

**Context (please always complete the following information):**
 - OS: macOS 11.1
 - eXist-db version: eXist 5.3.0-SNAPSHOT 35a76f25f0431220de95ff43752232bbf6a59efc 20210104165718
 - Java Version: 1.8.0_275

**Additional context**
- How is eXist-db installed? built from source
- Any custom changes in e.g. `conf.xml`? none

**Comment by joewiz** [comment]
I've updated the report to add an xqsuite version of the test.

**Comment by joewiz** [comment]
Here's how eXist's implementation of `fn:serialize` gets serialization parameters from conf.xml: https://github.com/eXist-db/exist/blob/adebc77a884fea5e0120c25290b9767ef01dce08/exist-core/src/main/java/org/exist/xquery/functions/fn/FunSerialize.java#L94-L106.

At least I think so. Here's `SerializerUtils.getSerializationOptions`: https://github.com/eXist-db/exist/blob/303ca0db511ee93bac1b50ccede357fde9658e7b/extensions/modules/file/src/main/java/org/exist/xquery/modules/file/SerializeToFile.java#L174-L196. Are the defaults for `getSerializationOptions` taken from conf.xml, or they defined in `ParameterConvention`? https://github.com/eXist-db/exist/blob/303ca0db511ee93bac1b50ccede357fde9658e7b/extensions/modules/file/src/main/java/org/exist/xquery/modules/file/SerializeToFile.java#L174-L196.

(I've also spotted hardcoded serialization parameters in `file:serialize`, though this function also allows you to override the hardcoded values: https://github.com/eXist-db/exist/blob/303ca0db511ee93bac1b50ccede357fde9658e7b/extensions/modules/file/src/main/java/org/exist/xquery/modules/file/SerializeToFile.java#L174-L196. It's puzzling why `file:serialize` needs to hardcode parameters at all. Shouldn't it just read from conf.xml?)

**Comment by line-o** [comment]
refs #4081

**Comment by joewiz** [comment]
@line-o I believe this was fixed by your #4254, wasn't it?

**Comment by line-o** [comment]
@joewiz I just tested it and it appears that settings in conf.xml do _not_ affect serialization of file:sync.

Digging a little deeper in the implementation of fn:serialize the settings in conf.xml are present in the Serializerpool instances. Borrowing one of them would do the trick.

https://github.com/eXist-db/exist/blob/adebc77a884fea5e0120c25290b9767ef01dce08/exist-core/src/main/java/org/exist/util/serializer/XQuerySerializer.java#L85-L103

**Comment by line-o** [comment]
This change is also considered as a breaking by @adamretter

**Comment by joewiz** [comment]
@line-o Ah, I see! Yes, with your PR, we can set serialization parameters, but file:sync just doesn't yet respect defaults in conf.xml. I can see why a new PR causing file:sync to respect conf.xml defaults could break existing code that didn't specify serialization params and would probably need to be in a major release. Thanks!

**Comment by adamretter** [comment]
Okay, so if I understand correctly. This should respect the `conf.xml` serializer settings. I also think @line-o mentioned that he can do that relatively easily. I also respect that doing so might be a **breaking change**. So if we can do that, let's schedule this for eXist-db 7.0.0. Does that suit everyone?

**Comment by joewiz** [comment]
@adamretter Yes, I just checked myself and performed the test in the report above with a build of eXist develop (eXist 6.1.0-SNAPSHOT 5e23335288da3f146981e335197efce51de5664e 20220719105055) whose conf.xml's serializer/@indent is set to "no". The file was synced to disk with indentation, i.e., as follows:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<foo>
    <bar/>
</foo>
```

This indicates that file:sync does not currently respect conf.xml. 

I agree that a fix would count as a breaking change, so the issue is correctly labeled as 7.0.0.

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

Status update after re-reading the current `Sync.java`:

**The "hard-coded serialization parameters" half is addressed.** Since **eXist-6.1.0** (commit 8506587, "allow to set serialization options in file:sync"), `file:sync`'s 3rd argument accepts a `map(*)` that, alongside `prune` / `after` / `excludes`, carries serialization options (`indent`, `omit-xml-declaration`, etc.) via `SerializerUtils.getSerializationOptions`. So callers can now override the serialization rather than being stuck with fixed values.

**The "doesn't respect `conf.xml` serializer" half is still open.** When no options map is given, the output properties come from a hard-coded `DEFAULT_PROPERTIES` in Java (`indent=yes`, `omit-xml-declaration=no`, `expand-xincludes=no`, `encoding=UTF-8`), not from the `conf.xml` serializer defaults. So with no map, `file:sync` still ignores `conf.xml`.

Suggested disposition: either close this as resolved-by-override (the substantive fix shipped in 6.1.0) with the `conf.xml`-defaults gap tracked as a smaller follow-up, or narrow the title to "`file:sync` ignores `conf.xml` serializer defaults" and keep it open as that low-priority gap. Happy to go whichever way the maintainers prefer.

(Separately, while exercising the options map we found a concrete bug in the override path — the `expand-xincludes` extension parameter can't be set to a non-default value through the map — which we'll file as its own issue with a minimal repro.)

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

One more defect in the same file:sync serialization area, worth recording here rather than as a separate issue: **the 6.1.0 options-map can't set the eXist extension parameter `expand-xincludes` to a non-default value.** `expand-xincludes=true` is silently ignored (output is serialized as if `expand-xincludes=no`); `=false` only appears to work because it matches the forced default.

Root cause is a key-form mismatch between the two halves of the option handling:

1. `SerializerUtils.getSerializationOptions(map)` reads the W3C parameters by their **string** key, but reads the eXist extension parameters (including `expand-xincludes`) only by their **QName** key — `Q{http://exist.sourceforge.net/NS/exist}expand-xincludes`. So a caller must supply `expand-xincludes` under the QName key, and `getSerializationOptions` records it correctly.
2. `Sync.getOptions` then runs a "fill defaults" loop that re-checks the **input** map by the **string** key:

   ```java
   for (String p : DEFAULT_PROPERTIES.stringPropertyNames()) {     // p = "expand-xincludes"
       if (optionsMap.get(new StringValue(this, p)).isEmpty()) {    // looks up the STRING key
           outputProperties.setProperty(p, DEFAULT_PROPERTIES.getProperty(p));  // overwrites with "no"
       }
   }
   ```

   Because the value was supplied under the **QName** key (the only form `getSerializationOptions` accepts), this string-key lookup always misses, and the loop overwrites the just-parsed value with the hard-coded default `expand-xincludes=no`.

Repro:

```xquery
xquery version "3.1";
import module namespace file = "http://exist-db.org/xquery/file";
let $_ := xmldb:store("/db", "target.xml", <p>included</p>)
let $_ := xmldb:store("/db", "host.xml",
    <doc xmlns:xi="http://www.w3.org/2001/XInclude"><xi:include href="target.xml"/></doc>)
return
    file:sync("/db", "/tmp/sync-out", map {
        QName("http://exist.sourceforge.net/NS/exist", "expand-xincludes") : true()
    })
(: expected: /tmp/sync-out/host.xml has the include expanded; actual: it is preserved unexpanded :)
```

Fix: the "fill defaults" loop must test whether the caller supplied each parameter using the **same key form** `SerializerUtils` accepts — the exist-namespace QName for the extension parameters — rather than the bare string key. This is separate from the conf.xml-defaults gap that is the main subject of this issue, but it's in the same family and the same function.
